DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/state_machine/README.md

```diff
@@ -0,0 +1,216 @@
+> - Questa*-Intel® FPGA Edition
+> - Questa*-Intel® FPGA Starter Edition
+> - ModelSim® SE
+>
+> When using the hardware compile flow, Intel® Quartus® Prime Pro Edition must be installed and accessible through your PATH.
+>
+> :warning: Make sure you add the device files associated with the FPGA that you are targeting to your Intel® Quartus® Prime installation.
+
+> **Note**: In oneAPI full systems, kernels that use SYCL Unified Shared Memory (USM) host allocations or USM shared allocations (and therefore the code in this tutorial) are only supported by Board Support Packages (BSPs) with USM support. Kernels that use these types of allocations can always be used to generate standalone IPs.
+
+## Key Implementation Details
+
+This code sample demonstrates two different implementations of a state machine using SYCL High-Level Synthesis (HLS) on Intel FPGAs: the naive version and the optimized version. We will compare and analyze the Quality of Result (QoR) differences between them.
+
+### Naive Implementation
+The naive implementation (`class NaiveStateMachine`) uses basic state machine logic with direct data read and write operations within a loop. Depending on the state, it either processes data in the `State::PROCESS` by multiplying it with a coefficient `coeff` or loads a new coefficient in `State::LD_COEFF` state. State transitions are manually handled within the loop using a swich-case statement, and computation/operations are coded in the body of the each case, then performed sequentially.
+
+### Optimized Implementation
+The optimized implementation (`class OptimizedStateMachine`) employs the `task_sequence` feature from the SYCL HLS compiler, which allows for more fine-grained control of parallelism and dependency management. Similar to the naive implementation, the state transition and states are defined in a switch-case statement with each case representing a state of the state machine. However, instead of defining the operation inside the case body, this approach defines tasks (`updateCoeff` and `compute`) that are executed asynchronously, enabling the isolation of long initiation intervals (II) and improving throughput by reducing dependency chains between states.
+
+### Limitations of the Naive Version
+
+The naive implementation suffers from several key limitations:
+- **Inefficient State Management:** State transitions and data processing are tightly coupled, leading to increased latency and reduced efficiency in state management.
+- **Dependency Bottlenecks:** Each state depends linearly on the completion of the previous state, creating bottlenecks and increasing the total execution time.
+
+![](assets/bottleneck.png)
+
+### Improvements brought by `task_sequence`
+
+Utilizing `task_sequence` in the optimized implementation offers significant enhancements:
+- **Enhanced Parallelism:** By decoupling computational dependencies, `task_sequence` allows for more parallel operations, improving overall execution speed and throughput.
```

Comment:
Can you explain what operations won't executing in parallel in naive design but were executing in parallel in optimized design?